File: elixir/allergies/lib/allergies.ex
defmodule Allergies do
  use Bitwise

  @allergies_bits %{
    "eggs" => 1,
    "peanuts" => 2,
    "shellfish" => 4,
    "strawberries" => 8,
    "tomatoes" => 16,
    "chocolate" => 32,
    "pollen" => 64,
    "cats" => 128
  }

  def list(flags) do
    @allergies_bits
    |> Map.keys
    |> Enum.filter(&allergic_to?(flags, &1))
  end

  def allergic_to?(flags, item) do
    (@allergies_bits[item] &&& flags) > 0
  end
end

File: haskell/allergies/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

import Data.Bits

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Eq, Show, Enum)

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo a n = testBit n (fromEnum a)

allergies :: Int -> [Allergen]
allergies n = filter (`isAllergicTo` n) [Eggs .. Cats]
